### PR TITLE
Move SnappBase logic into SnappBaseCore library

### PR DESCRIPTION
**Pull request by fleupold**
This PR moves all logic of the current SnapBase contract into a library, turning the contract itself into a thin wrapper and storage holder. For more information on solidity libraries check [here](https://solidity.readthedocs.io/en/v0.5.3/contracts.html#libraries).

This leads to a gas reduction for `SnappAuction.sol` deployment from 4.76M to 3.29M gas.

Unfortunately, I had to create a bunch of new view functions, since we cannot expose the library internal state as public (solidity won't let us). Otherwise the reduction would have been roughly twice as big. 
Nonetheless, this should give us some breathing room and we should probably do something similar for `SnappAuction.sol`

I realize and apologize that this change is quite large. It is copying a large chunk of contract code from one file into another and adjusts the tests at the same time. 
Please give it a careful review despite that, since it's easy to make a copy paste error on my side.

🤙

### CODE REVIEW

**Comment by fleupold** [comment]
If we were to turn off optimisation again, we'd go from 6.5M gas without the library to 4.5M with the library.

**Review by bh2smith — COMMENTED** [review]
It appears as if all the tests are passing, which would suggest that there were no copy paste errors. Am am unsure if the savings we have achieved are actually worth it. Do you have any form of comparison evidence that there is any savings here?

**Comment by fleupold** [comment]
> Do you have any form of comparison evidence that there is any savings here?

I'm not sure I understand what form of evidence you have in mind. I ran the migration once on current master and once on the feature branch and then compared gas costs for deployment.

We can see that we are saving 2M gas without the solc optimizer and still 1.5M gas with the optimizer.

**Comment by bh2smith** [comment]
Ok, I see now. Below are my calculations. 

```
Current Master without optimization - 6565544 (6.5 M) gas
Current Master with optimization - 4767238 (4.7 M) gas
```

```
This Branch without optimization - 4583358 (4.5 M) gas
This Branch with optimization  - 3291869 (3.2 M) gas
```

**Review by bh2smith — APPROVED** [review]
Yes, I agree this is a lot of code to go through, but I am a strong believer in our unit tests. I am very happy to see the gas savings with all functionality remaining the same. We should be cautious to adapt our event listener to listen for events coming from both `SnappBase` and `SnappAuction` since they are now being emitted from two distinct contracts.

File: contracts/libraries/SnappBaseCore.sol [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+     */
+    function openAccount(Data storage data, uint16 accountId) public {
+        require(accountId < MAX_ACCOUNT_ID, "Account index exceeds max");
+        require(
+            IdToAddressBiMap.insert(data.registeredAccounts, accountId, msg.sender),
+            "Address occupies slot or requested slot already taken"
+        );
+    }
+
+    function addToken(Data storage data, address _tokenAddress) public {
```
Comment (josojo) [inline contracts/libraries/SnappBaseCore.sol:93]: Having this big struct data implements everything very elegantly.

I am wondering, whether we have to buy this at a cost. Do we have to "pay" gas for calling the function with the whole struct although only two variables are used?

**Review by josojo — APPROVED** [review]
I checked the logic, looks all good. Good work!

Although I was shocked by the first glance at the pr, I think everything is still readable.
I would be interested in the gas comparison for function calls. How much more does it cost to call a function if the whole struct data is handed over each time compared to the original version?

**Comment by fleupold** [comment]
I ran `truffle test --g` on both current master and this branch.

Current Master:

```
·-----------------------------------------------------------------------------|----------------------------·
|                                     Gas                                     ·  Block limit: 8000000 gas  │
·················|··························|··········|··········|···········|··············|··············
|  Contract      ·  Method                  ·  Min     ·  Max     ·  Avg      ·  # calls     ·  usd (avg)  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  addToken                ·   72921  ·   87985  ·    80918  ·          68  ·       0.40  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  applyDeposits           ·   59435  ·   74499  ·    65130  ·           8  ·       0.32  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  applyWithdrawals        ·  792796  ·  809908  ·   800191  ·           7  ·       3.98  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  claimWithdrawal         ·       -  ·       -  ·    92459  ·           2  ·       0.46  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  deposit                 ·   73030  ·  159008  ·   137448  ·          32  ·       0.68  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  openAccount             ·   65383  ·   65447  ·    65408  ·          43  ·       0.33  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  requestWithdrawal       ·   52401  ·  138379  ·    59902  ·         114  ·       0.30  │
·-------------------------------------------|----------|----------|-----------|--------------|-------------·
```

This branch:
```
·-----------------------------------------------------------------------------|----------------------------·
|                                     Gas                                     ·  Block limit: 8000000 gas  │
············································|·································|·····························
|  Methods                                  ·           20 gwei/gas           ·       248.39 usd/eth       │
·················|··························|··········|··········|···········|··············|··············
|  Contract      ·  Method                  ·  Min     ·  Max     ·  Avg      ·  # calls     ·  usd (avg)  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  addToken                ·   74735  ·   89799  ·    82737  ·          68  ·       0.41  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  applyDeposits           ·   61292  ·   76356  ·    66987  ·           8  ·       0.33  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  applyWithdrawals        ·  794545  ·  811657  ·   801940  ·           7  ·       3.98  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  claimWithdrawal         ·       -  ·       -  ·    94760  ·           2  ·       0.47  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  deposit                 ·   74369  ·  160077  ·   138576  ·          32  ·       0.69  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  openAccount             ·   67172  ·   67236  ·    67197  ·          43  ·       0.33  │
·················|··························|··········|··········|···········|··············|··············
|  SnappBase     ·  requestWithdrawal       ·   53998  ·  139706  ·    61468  ·         114  ·       0.31  │
--------------------------------------------|----------|----------|-----------|--------------|--------------·
```

Costs have increase by ca. 1-2k gas.

**Comment by josojo** [comment]
Thanks for running the gas estimates. Great, 2k is not too much.

**Pull request merged**